optional-option-result.md

```diff
@@ -0,0 +1,126 @@
+- Topic Name: Optional Option/Result
+- Start Date: 2022-06-25
+- RFC PR: [vlang/rfcs#00000](https://github.com/vlang/rfcs/pull/00000)
+- V Issue: [vlang/v#00000](https://github.com/vlang/v/issues/00000)
+
+# Summary
+
+The language currently does not provide a way to let developers handle memory allocation issues.
+Many developers are happy with the state of things: memory allocation failures are rare and a panic is acceptable.
+However the language could (and probably should) provide a way to handle these issues.
+
+A way to achieve this goal would be to return option/result without requiring the caller to handle failure scenarios.
+
+# Motivation
+
+V aims to provide developers the tools required to write solid applications however panicing on memory allocation can leave the application and its data in a invalid state.
+
+The developers should be able to handle these border cases without causing undue complexity on simpler applications.
+
+# Guide-level explanation
+
+Looking at this code:
+```v
+
+fn new_thingy () {
+ return &Thingy{
+  buffer: []int{len: 10000, cap: 30000}
+ }
+}
+
+```
+
+The code could fail as the underlying `malloc` code could fail as array is calling malloc which can panic
+
+```v
+
+[unsafe]
+pub fn malloc(n isize) &u8 {
+ ...
+ if res == 0 {
+  panic('malloc($n) failed')
+ }
+ ...
+}
+
+```
+
+The syntax could be changed to allow a new operator ?? returning an option/result but leaving it up to the caller to handle it or not.
+
+```v
+
+fn new_thingy () ??{
+	return &Thingy{
+		buffer: []int{len: 10000, cap: 30000}
+	}
+}
+
+```
+
+```v
+
+fn new_unsafe_thingy() {
+  return new_thingy()
+}
+
+```
+
+```v
+
+fn new_safe_thingy() ? {
+  return new_thingy() or {
+    error('memory allocation failed')
+  }
+}
+
+```
+
+V may require an new "stricter" compilation option forcing developers to handle all possible optional option/result.
+
+It may be that `??` could be used when the absence of handling is fine and `!?` used when the handling of the `or` case must be performed when compiling using the stricter mode.
```

Comment:
:+1: (sounds like what I proposed above :wink:)

Reply:
Then, maybe `???` could mean `did you run my code or not`, and `????` could mean `no, really did it run` and `?????` could mean `come on, tell me already - did you run it or not`, and `??????` could mean...